### ISSUE TITLE
It seems dnfdragora doesn't respect repository priorities

### ISSUE THREAD

**Issue by L-U-T-i** [manatools/dnfdaemon#37]
It is very annoying for updates.

I've rebuilt Fedora packages for CentOS 8, and everything seems to work good for me, but list of updates.

I have EPEL repo with priority 50 and EPEL Playground repo with priority 55, because I prefer to have packages from EPEL (not Playground) installed.

"dnf update" shows no updates (as expected on already updated system), while I see plenty of packages to be updated (EPEL Playground packages over EPEL ones - with the same versions / releases...).

**Comment by Conan-Kudo** [comment]
the repo priorities thing comes from DNF itself, I'm confused why dnfdaemon wouldn't respect that, since we leave all that to DNF: https://dnf.readthedocs.io/en/latest/conf_ref.html#repo-priority-label

@j-mracek or @dmach, could you comment?

**Comment by L-U-T-i** [comment]
As written, DNF does respect it (doesn't offer me any updates on a freshly updated system), but something (can not judge weather it is dnf-daemon or dnfdragora, but it is manifested in dnfdragora - so I've filled the issue here) cause exactly the same versions (and releases) from EPEL Playground repo (with a priority 55 - so lower than EPEL repo with priority 50) to appear on the list of updates (to update the same packages of EPEL repo)...

**Comment by anaselli** [comment]
We need to investigate more here. I'm not sure i understood.
So priority should just give the priority to a repo when the same package is exposed, I mean IIUC if there is an update or a not installed package requested then dnf gives the one with lower number of priority value.

But the reporter seems to say that the system is updated.
That is odd, so probably the problem (or the feature) is really to address to dnfdragora, since for an old issue for which id did not show updates (or new packaged as updates) from external repos we added that an update is update or new package.

**Comment by anaselli** [comment]
No i can't get it. why dnf should not show anything if you have new updates that comes only from the lower priority repository? 

`If there is more than one candidate package for a particular operation, the one from a repo with the lowest priority value is picked, possibly despite being less convenient otherwise (e.g. by being a lower version).`

**more than one** candidate and not just **one** from a repo. Or do I have misunderstood anything?
I'm not saying we are wrong, i'm just trying to understand :)

**Comment by Conan-Kudo** [comment]
@L-U-T-i priorities do not affect searching and indexing packages, only selection of packages for transactions (e.g. installation).

**Comment by L-U-T-i** [comment]
That's what I get:
![Screenshot](https://user-images.githubusercontent.com/50832205/93369984-70194d80-f850-11ea-8341-ceb539217008.png)

I have for example amavis-2.12.0-9.el8.noarch.rpm from EPEL repo installed, but dnfdragora wants to update it to amavis-2.12.0-9.epel8.playground.noarch.rpm. Maybe because 9.epel8.playground is considered higher than 9.el8? But, it isthe same version/release, and priority settings as above helps me with dnf to not consider it as an update. So, I'd say dnfdaemon or dnfdragora does something wrong... Or, at least not exactly as dnf (just GUI), what is not OK anyway, or?

**Comment by L-U-T-i** [comment]
And, while making this screenshot, I've noticed dnfdragora window height at 2.1.0 can not be reduced below as on my screenshot (it would be convenient in my case to have it smaller - so bigger part of a terminal window would be visible...). Can this be reverted back to as in 2.0.4?

**Comment by anaselli** [comment]
Well that package is certainly from dnfdaemon, but you're right 2.12.0-9.el8 is lower than 2.12.0-9.epel8.playground.1 while Version is the same, Release (as you can see in your picture) is 9.epel8.playground.1, a versioning example [here](https://fedoraproject.org/wiki/Package_Versioning_Examples).
So that package is ready for update/upgrade for dnfdragora, don't know what to say more.

Concerning windows i assume you're using gtk one, so i'd expect such a limitation is due to libyui-gtk not dnfdragora, here below my Qt one:
![immagine](https://user-images.githubusercontent.com/1303486/93376393-d2c31700-f859-11ea-9187-ed405f123d3c.png)

**Comment by L-U-T-i** [comment]
I'm not sure if "9.el8" release shall be considered as lower than "9.epel8.playground"? "el8" and "epel8.playground" are added through a "dist" macro, so numerical part of release is in both cases just 9 (= equal).

If it would be as you say, also a "dnf update" command should provide the package upgrades, or?

In any case, I would expect to get exactly the same "results" (updated packages installed at the end) if I launch "dnf update" in shell / terminal or if I launch dnfdragora -2> updates -> select all...

BTW - I started to notice this behavior just recently (with dnfdragora 2.0.0). Older dnfdragora 1.1 didn't "push" EPEL Playground! Also the latest yumex-dnf doesn't, so I guess something went wrong with the last versions of dnfdragora...

I've made a test now with different versions of dnfdragora and just the latest dnfdaemon. So, looking at the results (screenshots) it is obvious it is dnfdragora >= 2 to be blamed by my opinion (dnfdaemon is the same also for yumex-dnf which performs as expected, as well as dnfdragora 1.1).

About the window height, I'm not sure I understand how libyui-gtk would force such a min. height for dnfdragora 2.1.0, but not for all other windows (as dnfdragora 2.0.4 / 2.0.0 / 1.1 or yumex-dnf)?! I suspect there is something else (as a min. number of rows visible or so) changed since dnfdragora 2.1.0 and 2.0.4...

Screenshots:
![Screenshot-1](https://user-images.githubusercontent.com/50832205/93399777-863e0280-f87e-11ea-93cb-c7e33181f305.png)

![Screenshot-2](https://user-images.githubusercontent.com/50832205/93399796-8f2ed400-f87e-11ea-8149-b05c097833c0.png)

![Screenshot-3](https://user-images.githubusercontent.com/50832205/93399806-93f38800-f87e-11ea-9971-d0c4f5e3718c.png)

![Screenshot-4](https://user-images.githubusercontent.com/50832205/93399817-98b83c00-f87e-11ea-9aa1-2066120c89aa.png)

**Comment by anaselli** [comment]
> I'm not sure if "9.el8" release shall be considered as lower than "9.epel8.playground"? "el8" and "epel8.playground" are added through a "dist" macro, so numerical part of release is in both cases just 9 (= equal).
> 

But it is. %dist macro is just for packaging is not for querying, At the time Mageia started upgrade from mandriva was supported and that  was reached also because package-Version-Release.mga1 was higher than package-Version-Release.mdvX.

> If it would be as you say, also a "dnf update" command should provide the package upgrades, or?
> 
> In any case, I would expect to get exactly the same "results" (updated packages installed at the end) if I launch "dnf update" in shell / terminal or if I launch dnfdragora 2 -> updates -> select all...
> 
Well i don't know what dnf does, i know what yumex-dnf was since i started from that code, but as i said in my previous comments i added also package to install as updates to fix an issue in which (and here i'm quite sure yumex-dnf failed at the time i started) dnfdragora didn't show updates from external repositories such as skype, mega etc. and that because they are often if not always new versions. So it's probably that the different behavior you noticed.
But i cannot evaluate if a pacakge is the same as the one installed or a new one if their releases are different, certainly both packager are different and built differently.
For instance in mageia we have two official repositories core and tainted, from the first you could get 

- avidemux-cli-2.7.6-2.mga8.x86_64.rpm

while from the latter 

- avidemux-cli-2.7.6-2.mga8.tainted.x86_64.rpm     
They are same release but one has different content, built differently etc... for instance not free code, patent etc..
As you can see mageia plays with release to grant upgrades, and users have it for free  by enabling the repo... and note that both repos are officially maintained by the distro.
So if i remove this feature i will be wrong for many users.
Maybe it could be disabled by an option that allows to show pure updates, but that requires to write code, i'm always ready to get patches from users of course :D

**Comment by anaselli** [comment]
> About the window height, I'm not sure I understand how libyui-gtk would force such a min. height for dnfdragora 2.1.0, but not for all other windows (as dnfdragora 2.0.4 / 2.0.0 / 1.1 or yumex-dnf)?! I suspect there is something else (as a min. number of rows visible or so) changed since dnfdragora 2.1.0 and 2.0.4...

Maybe you're right, but i don't seem to have used createMinSize for main dialog, i will check libyui-gtk behavior as soon as i can, but since it isn't manatool core but suse (should be really, since it does not provide changes there), i do that with lower priority, sorry.

**Comment by anaselli** [comment]
@L-U-T-i may ask you a test?
I made a search and i think i did that change [here](https://github.com/manatools/dnfdragora/pull/147/commits/bb0c85a39ff474fa0785c69f27c13623cff6d753#diff-6a15535e26dbe9e53bdd869800e5517aL1483).
So since i can't reproduce that now, i need to install fedora and add your repos, or find a test case here, you could do a test for me. I mean You could change a line into **dnfdragora/ui.py**.

```
diff --git a/dnfdragora/ui.py b/dnfdragora/ui.py
index bbf6d69..5497af3 100644
--- a/dnfdragora/ui.py
+++ b/dnfdragora/ui.py
@@ -1885,7 +1885,7 @@ class mainGui(dnfdragora.basedragora.BaseDragora):
                 # we requested installed for caching
                 self._populateCache('installed', po_list)
                 self.infobar.set_progress(0.33)
-                self._cachingRequest('updates_all')
+                self._cachingRequest('updates')
               elif self._status == DNFDragoraStatus.CACHING_UPDATE:
                 po_list = info['result']
                 # we requested updates for caching

```

Basically you have to change self._cachingRequest('updates_all') to self._cachingRequest('updates') and see if the problem you noticed still persists. If you experience your expected behaviour i could try to add an option to set and let the user to get one or the other behaviour. Default is updates_all of course, but that could do what you need anyway. WDYT?

**Comment by j-mracek** [comment]
I am sorry, I don't have much time to go into details how priority works. I suggest that the best way how to investigate the issue is to look at debugsolv data (not sure, how to get them from dnfdragora). I suggest that we will see that dnfdragora did some pre-selection before the data were given to the solver. Like if you have pkgA-1.0 (priority 50), pkgA-1.0.1 (priority 55) in transaction then dnf will pick pkgA-1.0 (priority 50). But in case that pkgA-1.0.1 (priority 55) will be recognized as a best candidate for the transaction by DNFdragora, then inserted in transaction as only candidate, then dnf will install pkgA-1.0.1 (priority 55), because it was requested. For proper functionality, the solver must see pkgA-1.0 (priority 50) even if it has the same version like installed in transaction, otherwise priority rule will be not applied.

**Comment by L-U-T-i** [comment]
anaselli, you're great - it is definitely this line. Changing as you suggest resolves the issue for me. Thank you.

An option (as you indicate) would be great, as one may prefer the way it is now. I definitely want exactly the same behavior as dnf itself provides.

About the window height - it seems to be independent of the desktop (screen) size, making dnfdragora more or less useless on smaller screens (buttons at the bottom are not visible / accessible). At least in VNC session (sorry, can not test on monitor right now...). So it may not be so small issue at the end.

By the way, would it be very complicated to make also the borders between panes movable? For me, the left pane (containing only "All" with an icon, about 10% or even less of pane width at full screen width of dnfdragora window...) is way too wide for me (unless select "Groups"), while the right one would be better if wider. And, the bottom one could most of the time be much lower...
